### PR TITLE
[stock][mrp] Usability

### PR DESCRIPTION
**Pull request by ryv-odoo**
[IMP] mrp: add unbuild button on MO

In this commit:
- On done MO's form, add an Unbuild button, it would open a
wizard with the form view of an unbuild, and some pre-filled fields
such as MO, product, BOM, quantity and total quantity finished which
can be edited.
- In the unbuild order, improve the on_change of mo_id:
if the selected product is tracked by serial number,
set quantity to 1 by default and readonly
- Also clean the testcase of unbuild order.

[IMP] stock, mrp, repair: improve the insufficient quantity wizards

Improve the inherited wizards of 'stock.warn.insufficient.qty' by adding
some piece of information about the quantity involved. Also
change titles to integrated the related product.

[IMP] stock: add an action to unreserve transfers in picking.

[IMP] stock: allow to search on reference for stock.production.lot

In this commit, we override the "name_search" method for search the
internal reference when searching on the name.

[IMP] product: set a quantity default hide on product form

In this commit, on the product form, in purchase tab, the quantity
field of the vendor pricelist should be set in optional = hide.

### CODE REVIEW

File: addons/mrp/views/mrp_unbuild_views.xml [hunk 1]
```diff
@@ -114,8 +114,8 @@
                                 <field name="bom_id" attrs="{'readonly':[('mo_id','!=',False)]}" force_save="1"/>
                                 <label for="product_qty"/>
                                 <div class="o_row">
-                                    <field name="product_qty"/>
-                                    <field name="product_uom_id" options="{'no_open':True,'no_create':True}" groups="uom.group_uom"/>
+                                    <field name="product_qty" attrs="{'readonly':[('has_tracking','=', 'serial')]}"/>
+                                    <field name="product_uom_id" options="{'no_open':True,'no_create':True}" groups="uom.group_uom" attrs="{'readonly':[('mo_id','!=',False)]}" force_save="1"/>
```
Comment (amoyaux) [inline addons/mrp/views/mrp_unbuild_views.xml]: missing spaces after
- :
- ,

File: addons/mrp/models/mrp_unbuild.py [hunk 2]
```diff
@@ -276,15 +276,16 @@ def action_validate(self):
             return self.action_unbuild()
         else:
             return {
-                'name': _('Insufficient Quantity'),
+                'name': self.product_id.display_name + _(' : Insufficient Quantity To Unbuild'),
```
Comment (amoyaux) [inline addons/mrp/models/mrp_unbuild.py]: remove the space before :

File: addons/repair/models/repair.py [hunk 3]
```diff
@@ -214,16 +214,17 @@ def action_validate(self):
                 return self.action_repair_confirm()
         else:
             return {
-                'name': _('Insufficient Quantity'),
+                'name': self.product_id.display_name + _(' : Insufficient Quantity To Repair'),
```
Comment (amoyaux) [inline addons/repair/models/repair.py]: the same

File: addons/stock/models/stock_scrap.py [hunk 4]
```diff
@@ -168,10 +168,11 @@ def action_validate(self):
             ctx.update({
                 'default_product_id': self.product_id.id,
                 'default_location_id': self.location_id.id,
-                'default_scrap_id': self.id
+                'default_scrap_id': self.id,
+                'default_quantity': self.scrap_qty
             })
             return {
-                'name': _('Insufficient Quantity'),
+                'name': self.product_id.display_name + _(' : Insufficient Quantity To Scrap'),
```
Comment (amoyaux) [inline addons/stock/models/stock_scrap.py]: the same

File: addons/repair/wizard/stock_warn_insufficient_qty_views.xml [hunk 5]
```diff
@@ -8,6 +8,9 @@
             <xpath expr="//field[@name='location_id']" position="after">
                 <field name="repair_id" invisible="True"/>
             </xpath>
+            <xpath expr="//div[@name='description']" position="after">
+                Do you confirm you want to repair <strong><field name="quantity" readonly="True"/></strong> units from location <strong><field name="location_id" readonly="True"/></strong>? This may lead to inconsistencies in your inventory.
```
Comment (Whenrow) [inline addons/repair/wizard/stock_warn_insufficient_qty_views.xml]: use product's uom

File: addons/stock/wizard/stock_warn_insufficient_qty_views.xml [hunk 6]
```diff
@@ -7,25 +7,24 @@
             <form>
                 <div>
                     The product
-                    <strong><field name="product_id" readonly="True"/></strong> is not available in sufficient quantity
+                    <strong><field name="product_id" readonly="True" invisible="1"/></strong> is not available in sufficient quantity
```
Comment (Whenrow) [inline addons/stock/wizard/stock_warn_insufficient_qty_views.xml]: remove the <fields/> entierely

File: addons/stock/wizard/stock_warn_insufficient_qty_views.xml [hunk 7]
```diff
@@ -7,25 +7,24 @@
...
                     <span class="oe_inline" groups="stock.group_stock_multi_locations"> in
                         <strong><field name="location_id" readonly="True"/></strong>.
                     </span>
                 </div>
                 <div attrs="{'invisible': [('quant_ids', '=', [])]}">
                     <br/>
-                    <strong>Here is your current inventory: </strong>
+                    <strong>Current Inventory Level: </strong>
```
Comment (Whenrow) [inline addons/stock/wizard/stock_warn_insufficient_qty_views.xml]: Current Inventory

File: addons/stock/wizard/stock_warn_insufficient_qty_views.xml [hunk 8]
```diff
@@ -44,6 +43,9 @@
             <xpath expr="//field[@name='product_id']" position="after">
                 <field name="scrap_id" invisible="1"/>
             </xpath>
+            <xpath expr="//div[@name='description']" position="inside">
+                Do you confirm you want to scrap <strong><field name="quantity" readonly="True"/></strong> units from location <strong><field name="location_id" readonly="True"/></strong>? This may lead to inconsistencies in your inventory.
```
Comment (Whenrow) [inline addons/stock/wizard/stock_warn_insufficient_qty_views.xml]: product's uom

File: addons/stock/models/stock_production_lot.py [hunk 9]
```diff
@@ -93,3 +94,13 @@ def action_lot_open_quants(self):
...
         return self.env['stock.quant']._get_quants_action()
+
+    @api.model
+    def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_uid=None):
+        """ search name and reference """
+        args = args or []
+        domain = []
+        if name:
+            domain = ['|', ('name', operator, name), ('ref', operator, name)]
+        move_ids = self._search(expression.AND([domain, args]), limit=limit, access_rights_uid=name_get_uid)
```
Comment (Whenrow) [inline addons/stock/models/stock_production_lot.py]: lot_ids

File: addons/stock/models/stock_production_lot.py [hunk 10]
```diff
@@ -93,3 +94,13 @@ def action_lot_open_quants(self):
...
+        return models.lazy_name_get(self.browse(move_ids).with_user(name_get_uid))
```
Comment (Whenrow) [inline addons/stock/models/stock_production_lot.py]: the same

File: addons/stock/models/stock_production_lot.py (hunk 9 above)
Comment (sle-odoo) [reply addons/stock/models/stock_production_lot.py]: @ryv-odoo did you ask lap we really want this feature ?

File: addons/stock/models/stock_scrap.py [hunk 11]
```diff
@@ -168,10 +168,11 @@ def action_validate(self):
             ctx.update({
                 'default_product_id': self.product_id.id,
                 'default_location_id': self.location_id.id,
-                'default_scrap_id': self.id
+                'default_scrap_id': self.id,
+                'default_quantity': self.scrap_qty
```
Comment (amoyaux) [inline addons/stock/models/stock_scrap.py]: scrap_qty

File: addons/repair/models/repair.py [hunk 12]
```diff
@@ -214,16 +214,17 @@ def action_validate(self):
                 return self.action_repair_confirm()
         else:
             return {
-                'name': _('Insufficient Quantity'),
+                'name': self.product_id.display_name + _(': Insufficient Quantity To Repair'),
```
Comment (amoyaux) [inline addons/repair/models/repair.py]: Fix the repair wizard. 1 unit in stock: repair 1 dozen -> No wizard

File: addons/mrp/models/mrp_unbuild.py [hunk 13]
```diff
@@ -276,15 +280,16 @@ def action_validate(self):
...
                 'view_mode': 'form',
                 'res_model': 'stock.warn.insufficient.qty.unbuild',
                 'view_id': self.env.ref('mrp.stock_warn_insufficient_qty_unbuild_form_view').id,
                 'type': 'ir.actions.act_window',
                 'context': {
                     'default_product_id': self.product_id.id,
                     'default_location_id': self.location_id.id,
-                    'default_unbuild_id': self.id
+                    'default_unbuild_id': self.id,
+                    'default_quantity': self.product_qty
```
Comment (amoyaux) [inline addons/mrp/models/mrp_unbuild.py]: Fix the unbuild wizard with UoM

File: addons/stock/wizard/stock_warn_insufficient_qty_views.xml [hunk 14]
```diff
@@ -6,26 +6,24 @@
...
                             <field name="location_id" options="{'no_create': True}"/>
                             <field name="lot_id" groups="stock.group_production_lot"/>
                             <field name="quantity"/>
                         </tree>
                     </field>
                 </div>
-                <div>
-                    Are you sure you want to confirm this operation? This may lead to inconsistencies in your inventory.
+                <div name="description">
+                    <!-- TO FILL by inherited views -->
```
Comment (sle-odoo) [inline addons/stock/wizard/stock_warn_insufficient_qty_views.xml]: remove the comment

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**